### ISSUE TITLE
Is there a plan to support dns query on SRV record?

### ISSUE THREAD

**Issue by johnlanni** [telepresenceio/telepresence#1863]
I found this code in `cmd/traffic/cmd/agent/client.go:120`.  It seems that telepresence's dns query only support the `A/AAAA` records, because use the `net.LookupHost` function.
```go
        go func() {
                for ctx.Err() == nil {
                        lr, err := lrStream.Recv()
                        if err != nil {
                                if ctx.Err() == nil {
                                        dlog.Debugf(ctx, "lookup request stream recv: %+v", err) // May be io.EOF
                                }
                                return
                        }
                        dlog.Debugf(ctx, "LookupRequest for %s", lr.Host)
                        addrs, err := net.LookupHost(lr.Host)
                        r := rpc.LookupHostResponse{}
                        if err == nil {
                                ips := make(iputil.IPs, len(addrs))
                                for i, addr := range addrs {
                                        ips[i] = iputil.Parse(addr)
                                }
                                dlog.Debugf(ctx, "Lookup response for %s -> %s", lr.Host, ips)
                                r.Ips = ips.BytesSlice()
                        }
                        response := rpc.LookupHostAgentResponse{
                                Session:  session,
                                Request:  lr,
                                Response: &r,
                        }
                        if _, err = manager.AgentLookupHostResponse(ctx, &response); err != nil {
                                if ctx.Err() == nil {
                                        dlog.Debugf(ctx, "lookup response: %+v %v", err, &response)
                                }
                                return
                        }
                }
        }()
```

But we use the `net.LookupSRV` function in our code. Since we want to use the dns SRV record to get the FQDN and port of the service through K8S service name.

Is there a plan to support this SRV query ?

**Comment by thallgren** [comment]
@johnlanni , there is no immediate plan to support SRV records but if there's a compelling reason to do so, then we might change that. Can you tell us more about your use case?

**Comment by johnlanni** [comment]
@thallgren , we have many micro services in our system, and every service has one or more independent ports. We use `K8S Service` for service discovery,  since the `K8S Service` mechanism will automatically create the corresponding SRV record in coredns. Because the complete service address can be obtained by the name of the service through SRV dns query, we don’t need to remember the port of each service.

This is an example service：
```yaml
apiVersion: v1
kind: Service
metadata:
  name: hepa
  namespace: default
spec:
  clusterIP: 10.99.217.177
  ports:
  - name: http
    port: 8080
    protocol: TCP
    targetPort: 8080
  - name: grpc
    port: 8081
    protocol: TCP
    targetPort: 8081
  - name: admin
    port: 8082
    protocol: TCP
    targetPort: 8082
  selector:
    erda/component: hepa
  sessionAffinity: None
  type: ClusterIP
```

This service has three ports:
- 8080: for http api
- 8081: for grpc api
- 8082: for admin api

When we want to access the hepa's grpc api, we will use the following code to get the full address.

```go
 _, addrs, err := net.DefaultResolver.LookupSRV(context.Background(), "grpc", "tcp", "hepa")
```

**Comment by dcd000** [comment]
Hi. We have the same problem. 
In deed for a Statefulset with a Headless Services the SRV type dns lookup is the primary way to access it, as the number of nodes may increase or decrease dynamically and the SRV recorsets keep track of the nodes.
Certainly this is a must if you want to work with statefulsets in kubernetes.

**Comment by onmomo** [comment]
@thallgren We run into a very similar issue as @johnlanni explained. We use the Akka GRPC client to load balance the GRPC request leveraging k8s headless service to lookup the replica pod IPs via SRV records as documented here.
* https://doc.akka.io/docs/akka/current/discovery/index.html#discovery-method-dns

```
nslookup -type=srv headless-service
```
won't return any records from my telepresence client only if executed directly in the cluster.
Would be awesome if telepresence could support this.

**Comment by thallgren** [comment]
Yes, there are plans to support SRV records (and others). I have some work pending on branch `thallgren/dns-experiments` that you might want to try out. Please note that this work is experimental and incomplete at this point. But perhaps trying it out can give an early indication of if it solves your problem?

**Comment by onmomo** [comment]
> thallgren/dns-experiments

Great! :-) Happy to test the branch. I tried to build it but it seems to miss some dns dependencies.
```
# github.com/telepresenceio/telepresence/rpc/v2/manager
TELEPRESENCE_VERSION=v2.7.1 make install
rpc/manager/manager_grpc.pb.go:104:37: undefined: DNSRequest
rpc/manager/manager_grpc.pb.go:104:76: undefined: DNSResponse
....
```

**Comment by thallgren** [comment]
@onmomo since it changes the RPC slightly, you'll need to start with a `make generate`.

**Comment by thallgren** [comment]
@onmomo also, I think you should use another version, such as `v2.7.2-alpha.0` or similar. You will need to reinstall the traffic-manager (because of the RPC change).

**Comment by onmomo** [comment]
@thallgren ok thanks, tried again. Looks like the MacOS build has issues now.
```
 TELEPRESENCE_VERSION=v2.7.2-alpha.0 make generate install
# github.com/telepresenceio/telepresence/v2/pkg/client/rootd/dns
pkg/client/rootd/dns/server_darwin.go:182:73: not enough arguments in call to s.Run
        have (context.Context, chan struct{}, []net.PacketConn, nil, func(c context.Context, q *"github.com/miekg/dns".Question) (result []"github.com/miekg/dns".RR, err error))
        want (context.Context, chan<- struct{}, *listeners, FallbackPool, FallbackPool, Resolver)
```
setting the missing argument fallbackPoolTCP to nil, led to
```
# github.com/telepresenceio/telepresence/v2/pkg/client/rootd/dns
pkg/client/rootd/dns/server_darwin.go:182:40: cannot use []net.PacketConn{…} (value of type []net.PacketConn) as type *listeners in argument to s.Run
```
I need to give up solving this type problem, I'm way too far away from Go it seems. :-)

**Comment by thallgren** [comment]
Ok, I'll take a look later today. Seems I left things in a bad state. Sorry.

**Comment by thallgren** [comment]
@onmomo should be OK now. I rebased it and did a `make generate` so you should be able to build from what's there now.

**Comment by onmomo** [comment]
> @onmomo should be OK now. I rebased it and did a `make generate` so you should be able to build from what's there now.

was able to `make generate build` and test it now.

**Comment by onmomo** [comment]
@thallgren I managed to build and push the image. (M1 fix -> https://github.com/telepresenceio/telepresence/pull/2759). Then uninstalled the traffic manager and installed again via telepresence helm install.
Unfortunately, it behaves the same as before. The pod IP can't be resolved via the headless service.

**Comment by onmomo** [comment]
I don't really see any SRV LookupDNS log entries. Looks like the request is not really forwarded to traffic-manager
```
WARNING: [Channel<1>: (//headless-svc.namespace)] Failed to resolve name. status=Status{code=UNAVAILABLE, description=NameResolver returned no usable address. addrs=[], attrs={}
akka.grpc.internal.AkkaDiscoveryNameResolver@4e4c2b8e was used, cause=null}
```

I would expect if I run this on the terminal, that I would see this logged in telepresence?
```
nslookup -type=srv headless-svc.namespace
Server:		1.1.1.1
Address:	1.1.1.1#53

** server can't find headless-svc.namespace: NXDOMAIN
```

**Comment by thallgren** [comment]
Try `telepresence loglevel debug`, then try the `nslookup` again and check the `daemon.log`. Does the DNS lookup show up there?

**Comment by onmomo** [comment]
> Try `telepresence loglevel debug`, then try the `nslookup` again and check the `daemon.log`. Does the DNS lookup show up there?

There is actually nothing in the daemon.log referring to the nslookup request. I also debugged the JVM code and figured out that the dns lookup requests `headless-svc.namespace.$searchDomain`. $SearchDomain matches the domain defined in the MacOS DNS settings for "Search Domains". Not sure why the JVM resolves this address for the DNS lookup or what Search Domains actually is used for but this seems to be the reason why telepresence does not forward it to the cluster.
I can try to add this domain to the `include-suffixes` to see if it would catch it then but this has probably other implications.

<img width="624" alt="Screenshot 2022-08-26 at 09 17 44" src="https://user-images.githubusercontent.com/1713109/186847239-272d9855-69e6-43fb-ae5f-53bf4a36af87.png">

**Comment by thallgren** [comment]
There was a glitch in how the cache worked. I fixed and pushed a fix for it. When doing this:
```console
$ telepresence connect
$ telepresence loglevel debug
$ host -t srv _myservice._tcp.tada.se
```
This is what you should expect to see in the logs (I'm using my own domain in these samples, because I added a server record there)

In the daemon log:
```
09:52:08.7720 debug   daemon/session/dns/resolved/Server : Lookup SRV "_myservice._tcp.tada.se."
09:52:08.8310 debug   daemon/session/dns/resolved/Server : SRV    _myservice._tcp.tada.se. -> NOERROR _myservice._tcp.tada.se.	4	IN	SRV	10 5 448 tada.se.
```

In the log of the `traffic-manager` pod:
```
2022-08-26 07:51:13.4670 debug   Logging at this level "debug"
2022-08-26 07:52:11.7441 debug   httpd/conn=127.0.0.1:8081 : LookupDNS SRV _myservice._tcp.tada.se. : session_id="a2d64c0f-7652-4195-abd9-dc92253e8a61"
2022-08-26 07:52:11.8017 debug   httpd/conn=127.0.0.1:8081 : LookupDNS returns [_myservice._tcp.tada.se.	4	IN	SRV	10 5 448 tada.se.] : session_id="a2d64c0f-7652-4195-abd9-dc92253e8a61"
```

**Comment by thallgren** [comment]
Oh, and the output of the last `host` command should be:
```
_myservice._tcp.tada.se has SRV record 10 5 448 tada.se.
```

**Comment by thallgren** [comment]
@onmomo forgot to mention. You will of course need to use the `include-suffixes` to run my particular example. I wouldn't expect "tada.se" to get routed to your DNS :-)

**Comment by onmomo** [comment]
@thallgren Ok, I integrated the cache fixed into my build and tested the provided host command.
I tested first without `include-suffixes` set to `tada.se` and I received the SRV response from the `host` command. Then I added the `include-suffixes` with the same result.
But it seems not to be resolved via telepresence, since there aren't any debug logs for this particular lookup only for others.
I also noticed that the JVM app does the lookup only via the OS configured 1.1.1.1 DNS, the same it seems for `host` and `nslookup` regardless what is configured on `include-suffixes`

The JVM app tries to resolve the k8s headless service also via 1.1.1.1 with and without the configured search domains in place.  
```
Resolving [_grpc._tcp.headless-svc.namespace] (SRV)
Message to [/1.1.1.1:53]: [Message(435,<QUERY,RD,SUCCESS>,List(Question(_grpc._tcp.headless-svc.namespace,RecordType(33,SRV),RecordClass(1,IN))),List(),List(),List())]
```

**Comment by thallgren** [comment]
@onmomo must be some macOS specific behavior then (I'm mostly on Linux). I won't be able to look into that until Tuesday next week at the earliest, because I'm enjoying PTO.

**Comment by onmomo** [comment]
> @onmomo must be some macOS specific behavior then (I'm mostly on Linux). I won't be able to look into that until Tuesday next week at the earliest, because I'm enjoying PTO.

@thallgren yes, it seems. Looks like there is already a [ticket](https://github.com/telepresenceio/telepresence/issues/2682). I wonder if there is a workaround for the JVM. Enjoy your time off and thanks for the support!

**Comment by thallgren** [comment]
@onmomo please note that that ticket won't solve your problem. Using `/etc/resolver` is a supported way to resolve on macOS. Using `nslookup` however, isn't really. From the nslookup man-page:
```
   The nslookup command does not use the host name and address resolution or the DNS 
   query routing mechanisms used by other processes running on Mac OS X.  The results of 
   name or address queries printed by nslookup may differ from those found by other
   processes that use the Mac OS X native name and address resolution mechanisms. The 
   results of DNS queries may also differ from queries that use the Mac OS X DNS routing 
   library.
```
The same is true about some other commands. `dig` most notably, and also `host` which I used in my example.
On a macOS, you can try using `dns-sd -q _grpc._tcp.headless-svc.namespace SRV`

**Comment by onmomo** [comment]
@thallgren thanks a lot for figuring this out. Providing the **full service name**, it works but only sometimes most likely due to the following error reported in the daemon logs.
working response
```
dns-sd -q _grpc._tcp.headless-svc.namespace.svc.cluster.local SRV
DATE: ---Mon 29 Aug 2022---
 9:46:49.179  ...STARTING...
Timestamp     A/R    Flags if Name                          Type  Class   Rdata
 9:46:49.216  Add        2  0 _grpc._tcp.headless-svc.namespace.svc.cluster.local.SRV    IN     0 100 3000 10-65-116-89.headless-svc.namespace.svc.cluster.local.

```

not working
```
dns-sd -q _grpc._tcp.headless-svc.namespace.svc.cluster.local SRV
10:00:44.158  Add        2  0 _grpc._tcp.headless-svc.namespace.svc.cluster.local    IN     0.0.0.0    No Such Record

daemon/session/dns/Server : SRV    _grpc._tcp.headless-svc.namespace.svc.cluster.local. -> SERVFAIL rpc error: code = Internal desc = unable to pack DNS reply: dns: buffer size too small
OR
debug   daemon/session/dns/Server : SRV    _grpc._tcp.headless-svc.namespace.svc.cluster.local. -> SERVFAIL rpc error: code = Internal desc = unable to pack DNS reply: dns: overflow packing uint16
```

Is it a requirement to provide the full service name? Providing just the namespace always leads to
```
daemon/session/dns/Server : SRV    _grpc._tcp.headless-svc.namespace.. -> SERVFAIL rpc error: code = Internal desc = lookup _grpc._tcp.headless-svc.namespace. on 172.20.0.10:53: no such host
```
I can test again with the JVM and the full service name

**Comment by thallgren** [comment]
@onmomo I just pushed a bunch of improvements to the thallgren/dns-experiments branch. I think they should solve most of your issues. If not, please let me know.

**Comment by thallgren** [comment]
Closing this, because it's implemented in #2787.
